### ISSUE TITLE
type checking of member assigned with lambda

### ISSUE THREAD

**Issue by beauxq** [microsoft/pyright#4011]
## Code Snippet

```python
from typing import Callable


class C:
    f: Callable[[int], bool] = staticmethod(lambda x: print(type(x)) or True)


c = C()
c.f(3)


class D:
    f: Callable[[int], bool] = lambda x: print(type(x)) or True


d = D()
d.f(3)
```

## Expected behavior

The one that works correctly should not have a pylance error.
The one that doesn't work correctly should be found by pylance type checking.

## Actual behavior

The first one, class C, works correctly, but pylance says it's wrong.
output:
```
<class 'int'>
```
pylance:
`Expression of type "staticmethod[bool]" cannot be assigned to declared type "(int) -> bool"`

The second one, class D, fails, but pylance doesn't see any problem with it.
output:
```
Traceback (most recent call last):
  File ...
    d.f(3)
TypeError: <lambda>() takes 1 positional argument but 2 were given
```

**Comment by erictraut** [comment]
You said that pylance indicates that there is an error here. I'm not seeing any error reported. Can you confirm that you are seeing an error with the posted sample? Perhaps you simplified the sample after initially observing an error?

There are a few things going on here. You're bumping up against some limitations and ambiguities of the Python type system.

First, there is no way in the Python type system to explicitly specify whether a `Callable` type should be interpreted as an instance method (i.e. it requires binding to an object), a class method (i.e. it requires binding to a class), or a static method or normal function (i.e. requires no binding). Because there's no way to explicitly specify this in a `Callable` type annotation, a type checker will generally interpret it as a normal function (i.e. requires no binding).

Second, none of the Python typing standards are very clear about how class-scoped variable declarations should be interpreted — i.e. whether they should be treated as instance-only variables, class-only variables, or class variables that can be overwritten with an instance variable of the same type. Pyright's interpretation is documented [here](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#class-and-instance-variables).

In your example above, the class-scoped symbol `f` is declared as a "regular class variable" (which means it can be overwritten by an instance variable of the same type).

Consider the following variation on your `D` class definition:
```python
class D:
    def __init__(self):
        self.f = lambda x: print(type(x)) or True

    f: Callable[[int], bool]

d = D()
d.f(3)
```

The above sample type checks and runs without error.

Third, you are using `staticmethod` as a general function call rather than as a decorator, as it is normally used. Type checkers like pyright and mypy special case `staticmethod` (and `classmethod`) when they are used as a decorator. When used in other places, it is treated just like any other function call — or in this case, as a constructor call because `staticmethod` is declared as a class. So the call you're making in this case results in a type of `staticmethod[bool]`. The input parameter information is unfortunately lost in this type evaluation because the `staticmethod` constructor uses a `Callable[..., R]` type form. That means a type checker will not be able to enforce the input parameter types in the statement `f: Callable[[int], bool] = staticmethod(lambda x: print(type(x)) or True)`.

I'll think about this a bit more, but I don't currently see any way to achieve the expected behavior you've suggested — at least without creating false positive errors in more common use cases.

Depending on your intent with this code pattern, there are likely ways to change your code such that it can be type checked more thoroughly.

I'll note that mypy also does not emit any error for the code sample at the top of this thread. I see that you posted a [related bug report](https://github.com/python/mypy/issues/13818) in the mypy issue tracker.

**Comment by beauxq** [comment]
![pylance-function-property](https://user-images.githubusercontent.com/6052147/194141182-465c8b65-7382-4efb-a730-ef6827c19efa.png)
![mypy-function-property](https://user-images.githubusercontent.com/6052147/194141197-ff1630cf-05a1-42bf-99da-43c44d3f896a.png)

I don't like the usage of `staticmethod` because that name doesn't reflect how I'm thinking of this variable.
I think the typing `Callable[[int], bool]` does reflect well how I'm thinking of this variable.
It is a funciton that is stored on this member that can be changed on instances at runtime, with a default set by the class.
The `staticmethod` is just the way that someone found to make it work.

I see this as safer than:
```python3
class C:
    f: Callable[[int], bool]

    def __init__(self) -> None:
        self.f = lambda x: print(type(x)) or True


c = C()
c.f(3)
```

because someone subclassing this could forget to call `super().__init__()`, and then the default wouldn't be there.
But I think it might be good enough as a workaround.

I could see it as a problem with Python, rather than a problem with the type checkers, but if it were fixed in Python, it would break systems that exist in the wild.

I recognize this is a complex situation.
As with most type-checking, I see the false-negative as worse than the false-positive.

**Comment by erictraut** [comment]
Ah, I'm not seeing an error because I'm using Python 3.10. You must be using Python 3.9 or earlier. The `staticmethod` class definition in typestub defines a `__call__` method for Python 3.10 and newer. This makes it compatible with `Callable`.

```python
class staticmethod(Generic[_R_co]):
    ...
    if sys.version_info >= (3, 10):
        ...
        def __call__(self, *args: Any, **kwargs: Any) -> _R_co: ...
```

> As with most type-checking, I see the false-negative as worse than the false-positive.

That's the opposite of what most people expect with type checkers and contrary to the general philosophy we adopt with pyright. Keep in mind that there will always be classes of bugs that a static type checker will not be able to catch.

**Comment by beauxq** [comment]
> That's the opposite of what most people expect with type checkers

Do you have some data you've collected to back that up?

Keep in mind that the statement you replied to is only about the classes of bugs that a static type checker can catch.

**Comment by erictraut** [comment]
You've reported two issues. The first is a false positive. I think pyright is doing the right thing here based on the type information provided. The error goes away if you use Python 3.10 or newer because of the `__call__` method in the `staticmethod` class.

For the second issue, I agree this is a false negative, but I don't see a good way to address it without adding a bunch of logic that would be very specific to this code pattern. We generally don't add such case-specific logic unless it's a really common coding pattern. This doesn't come close to meeting that bar.

If your intent is for `f` to be an instance variable, I recommend using the workaround that I supplied above where you assign it in the constructor.
